### PR TITLE
Multiple crops

### PR DESCRIPTION
**Pull request by naushir**

### CODE REVIEW

**Comment by naushir** [comment]
@davidplowman my torture tests are still running, likely only done by the weekend now.  However I expect everything to pass now, so we ought to get ready to merge this soon.

Given the magnitude of the changes, can you check through my work please?

File: src/libpisp/backend/tiling/crop_stage.cpp [hunk 1]
```diff
@@ -13,7 +13,7 @@
 using namespace tiling;
 
 CropStage::CropStage(char const *name, Stage *upstream, Config const &config, int struct_offset)
-	: BasicStage(name, upstream->GetPipeline(), upstream, struct_offset), config_(config)
+	: BasicStage(name, upstream->GetPipeline(), upstream, struct_offset), config_(config), started_(false)
```
Comment (davidplowman) [inline src/libpisp/backend/tiling/crop_stage.cpp]: Are we using the `started_` field? I wasn't sure...

Comment (davidplowman) [reply src/libpisp/backend/tiling/crop_stage.cpp]: Ah no, I see we are using it! Please ignore me...

File: src/libpisp/backend/tiling/crop_stage.cpp [hunk 2]
```diff
@@ -72,3 +97,8 @@ void CropStage::PushCropDown(Interval interval, Dir dir)
 	PISP_LOG(debug, "(" << name_ << ") Exit with interval " << output_interval_);
 	downstream_->PushCropDown(output_interval_, dir);
 }
+
+void CropStage::Reset()
+{
```
Comment (davidplowman) [inline src/libpisp/backend/tiling/crop_stage.cpp:125]: Should we call the parent `BasicStage::Reset()`? I'm guessing it makes no difference...

File: src/libpisp/backend/tiling/crop_stage.cpp [hunk 3]
```diff
@@ -98,7 +121,7 @@ void CropStage::PushCropDown(Interval interval, Dir dir)
 	downstream_->PushCropDown(output_interval_, dir);
 }
 
-void CropStage::Reset()
+bool CropStage::BranchInactive() const
 {
-	started_ = false;
+	return !output_interval_.length;
```
Comment (davidplowman) [inline src/libpisp/backend/tiling/crop_stage.cpp:126]: I guess I feel slightly weird that `BranchInactive()` looks at the output_interval, but `BranchComplete()` doesn't - it slight offends my need for symmetry. Should `BranchCompletel()` look at `output_interval_.End()` or something? Or do I worry too much?

File: src/libpisp/backend/tiling/crop_stage.cpp (hunk 1 above)
Comment (davidplowman) [reply src/libpisp/backend/tiling/crop_stage.cpp]: Ah yes, the next commit removes it. That's probably what I remembered. Just getting confused!

File: src/libpisp/backend/tiling/rescale_stage.hpp [hunk 4]
```diff
@@ -46,7 +46,7 @@ class RescaleStage : public BasicStage
 
 private:
 	Config config_;
-	uint32_t round_up;
```
Comment (davidplowman) [inline src/libpisp/backend/tiling/rescale_stage.hpp:49]: Ouch! That must have reacted badly to the negative offsets we now have...

**Comment by davidplowman** [comment]
Mostly this all looks pretty good to me. I guess I've seen most of it before. Though there are surprisingly few changes compared to the amount of head-scratching required to find them!

I'm still wondering a little bit about the symmetry (or lack of) between `BranchComplete()` and `BranchInactive()`. If you wanted you could leave it with me and I could scratch my head some more on that question?

**Comment by naushir** [comment]
> Mostly this all looks pretty good to me. I guess I've seen most of it before. Though there are surprisingly few changes compared to the amount of head-scratching required to find them!
> 
> I'm still wondering a little bit about the symmetry (or lack of) between `BranchComplete()` and `BranchInactive()`. If you wanted you could leave it with me and I could scratch my head some more on that question?

Yes, perhaps we can replace this with ``output_interval_.End()``?  Will have to rerun all the tests again...

**Comment by naushir** [comment]
@davidplowman https://github.com/raspberrypi/libpisp/pull/32/commits/3a61f12f42e0d5d18323ccbd690e84e88a931ca0 makes things neater with the branch complete flag.  I think this is now good to go.

**Pull request merged**